src/core/util/hugepage_mgr.cpp

```diff
@@ -0,0 +1,258 @@
+{
+    std::vector<size_t> hugepages;
+    size_t best_hugepage = 0;
+
+    if (safe_mce_sys().hugepage_log2 == 0) {
+        get_supported_hugepages(hugepages);
+        std::sort(hugepages.begin(), hugepages.end(), std::greater<size_t>());
+    } else {
+        // User requested specific hugepage size - don't check other types.
+        hugepages.push_back(1LU << safe_mce_sys().hugepage_log2);
+    }
+
+    // This is naive algorithm and may work inefficiently in complex scenarios.
+    for (size_t hugepage : hugepages) {
+        if (get_free_hugepages(hugepage) * hugepage >= size) {
+            if (is_hugepage_optimal(hugepage, size)) {
+                return hugepage;
+            }
+            if (is_hugepage_acceptable(hugepage, size) && best_hugepage == 0) {
+                best_hugepage = hugepage;
+            }
+        }
+    }
+
+    if (best_hugepage &&
+        hugepage_unused_space(best_hugepage, size) > HUGEPAGE_UNUSED_WARNING_THRESHOLD) {
+        __log_info_dbg("Allocating %zu bytes with hugepages %zu kB.", size, best_hugepage / 1024U);
+        __log_info_dbg("%zu bytes of the last hugepage may be unused.",
+                       best_hugepage - size % best_hugepage);
+    }
+    return best_hugepage;
+}
+
+void *hugepage_mgr::alloc_hugepages(size_t &size)
+{
+    std::lock_guard<decltype(m_lock)> lock(m_lock);
+
+    size_t hugepage = find_optimal_hugepage(size);
+    size_t hugepage_mask = hugepage - 1;
+    size_t actual_size = size;
+    void *ptr = nullptr;
+
+    if (hugepage) {
+        int map_extra = 0;
+
+        if (hugepage != m_default_hugepage) {
+            map_extra = (int)log2(hugepage) << MAP_HUGE_SHIFT;
+        }
+        actual_size = (size + hugepage_mask) & ~hugepage_mask;
+
+        __log_info_dbg("Allocating %zu bytes with hugepages %zu kB", actual_size, hugepage / 1024U);
+
+        ptr = mmap(NULL, actual_size, PROT_READ | PROT_WRITE,
+                   MAP_PRIVATE | MAP_ANONYMOUS | MAP_POPULATE | MAP_HUGETLB | map_extra, -1, 0);
+        if (ptr == MAP_FAILED) {
+            ptr = nullptr;
+            __log_info_dbg("mmap failed (errno=%d)", errno);
+        } else {
+            m_hugepages[hugepage].nr_hugepages_free -= actual_size / hugepage;
+        }
+    }
+
+    // Statistics
+    m_stats.total_requested += actual_size;
+    if (ptr) {
+        m_hugepages[hugepage].nr_hugepages_allocated += actual_size / hugepage;
+        ++m_hugepages[hugepage].nr_allocations;
+        ++m_stats.allocations;
+        m_stats.total_allocated += actual_size;
+        m_stats.total_unused += actual_size - size;
+
+        size = actual_size;
+    } else {
+        ++m_stats.fails;
+    }
+    return ptr;
+}
+
+void hugepage_mgr::dealloc_hugepages(void *ptr, size_t size)
+{
+    int rc = munmap(ptr, size);
+    if (rc != 0) {
+        __log_info_dbg("munmap failed (errno=%d)", errno);
+    }
+    // We don't track released hugepages. Usually they're used for preallocation
```

Comment:
XLIO specific?

Reply:
Yes, xlio_allocator is not a general purpose allocator. Without tracking released pages we don't have to lock this method and this also simplifies printing a report at XLIO termination (a future feature).